File: app/src/com/tadaoyamaoka/vocalpitchmonitor/MathEval.java
package com.tadaoyamaoka.vocalpitchmonitor;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * A small arithmetic expression evaluator that mirrors the JavaScript
 * {@code eval(...)} semantics used by musescore-xen-tuner's tuning-config
 * parser for cents/ratio values.
 *
 * Supported syntax:
 *  - numbers (integer, decimal, exponent), parentheses, unary minus
 *  - + - * / with JS precedence
 *  - exponentiation: Math.pow(a,b) and also ** / ^ (convenience)
 *  - Math.* / MATH.* function calls: pow, log, log2, log10, exp, sqrt, cbrt,
 *    abs, floor, ceil, round, trunc, sign, min, max, sin, cos, tan, asin,
 *    acos, atan, atan2, sinh, cosh, tanh, hypot
 *  - Math.* constants: PI, E, LN2, LN10, LOG2E, LOG10E, SQRT2, SQRT1_2
 *    (bare PI / E also accepted as a convenience)
 */
public class MathEval {

    public static final double NaN = Double.NaN;

    private interface Fn {
        double apply(double[] args);
    }

    private static final Map<String, Fn> FUNCS = new HashMap<String, Fn>();
    private static final Map<String, Double> CONSTS = new HashMap<String, Double>();

    static {
        FUNCS.put("pow", new Fn() { public double apply(double[] a) { return Math.pow(a[0], a[1]); } });
        FUNCS.put("log", new Fn() { public double apply(double[] a) { return Math.log(a[0]); } });
        FUNCS.put("log2", new Fn() { public double apply(double[] a) { return Math.log(a[0]) / Math.log(2.0); } });
        FUNCS.put("log10", new Fn() { public double apply(double[] a) { return Math.log10(a[0]); } });
        FUNCS.put("ln", new Fn() { public double apply(double[] a) { return Math.log(a[0]); } });
        FUNCS.put("exp", new Fn() { public double apply(double[] a) { return Math.exp(a[0]); } });
        FUNCS.put("sqrt", new Fn() { public double apply(double[] a) { return Math.sqrt(a[0]); } });
        FUNCS.put("cbrt", new Fn() { public double apply(double[] a) { return Math.cbrt(a[0]); } });
        FUNCS.put("abs", new Fn() { public double apply(double[] a) { return Math.abs(a[0]); } });
        FUNCS.put("floor", new Fn() { public double apply(double[] a) { return Math.floor(a[0]); } });
        FUNCS.put("ceil", new Fn() { public double apply(double[] a) { return Math.ceil(a[0]); } });
        FUNCS.put("round", new Fn() { public double apply(double[] a) { return Math.round(a[0]); } });
        FUNCS.put("trunc", new Fn() { public double apply(double[] a) { return a[0] < 0 ? Math.ceil(a[0]) : Math.floor(a[0]); } });
        FUNCS.put("sign", new Fn() { public double apply(double[] a) { return Math.signum(a[0]); } });
        FUNCS.put("min", new Fn() { public double apply(double[] a) { return Math.min(a[0], a[1]); } });
        FUNCS.put("max", new Fn() { public double apply(double[] a) { return Math.max(a[0], a[1]); } });
        FUNCS.put("sin", new Fn() { public double apply(double[] a) { return Math.sin(a[0]); } });
        FUNCS.put("cos", new Fn() { public double apply(double[] a) { return Math.cos(a[0]); } });
        FUNCS.put("tan", new Fn() { public double apply(double[] a) { return Math.tan(a[0]); } });
        FUNCS.put("asin", new Fn() { public double apply(double[] a) { return Math.asin(a[0]); } });
        FUNCS.put("acos", new Fn() { public double apply(double[] a) { return Math.acos(a[0]); } });
        FUNCS.put("atan", new Fn() { public double apply(double[] a) { return Math.atan(a[0]); } });
        FUNCS.put("atan2", new Fn() { public double apply(double[] a) { return Math.atan2(a[0], a[1]); } });
        FUNCS.put("sinh", new Fn() { public double apply(double[] a) { return Math.sinh(a[0]); } });
        FUNCS.put("cosh", new Fn() { public double apply(double[] a) { return Math.cosh(a[0]); } });
        FUNCS.put("tanh", new Fn() { public double apply(double[] a) { return Math.tanh(a[0]); } });
        FUNCS.put("hypot", new Fn() { public double apply(double[] a) { return Math.hypot(a[0], a[1]); } });

        CONSTS.put("PI", Math.PI);
        CONSTS.put("E", Math.E);
        CONSTS.put("LN2", Math.log(2.0));
        CONSTS.put("LN10", Math.log(10.0));
        CONSTS.put("LOG2E", 1.0 / Math.log(2.0));
        CONSTS.put("LOG10E", 1.0 / Math.log(10.0));
        CONSTS.put("SQRT2", Math.sqrt(2.0));
        CONSTS.put("SQRT1_2", 1.0 / Math.sqrt(2.0));
    }

    private String src;
    private int pos;

    private MathEval(String src) {
        this.src = src;
        this.pos = 0;
    }

    /**
     * Evaluate a JS-like arithmetic expression.
     * @return the numeric result, or NaN if the expression is invalid.
     */
    public static double eval(String expression) {
        if (expression == null) {
            return NaN;
        }
        String s = expression.trim();
        if (s.length() == 0) {
            return NaN;
        }
        try {
            MathEval p = new MathEval(s);
            double v = p.parseExpression();
            p.skipWs();
            if (p.pos < p.src.length()) {
                return NaN; // trailing garbage
            }
            return v;
        } catch (Exception e) {
            return NaN;
        }
    }

    private void skipWs() {
        while (pos < src.length()) {
            char c = src.charAt(pos);
            if (c == ' ' || c == '\t' || c == '\n' || c == '\r') {
                pos++;
            } else {
                break;
            }
        }
    }

    private double parseExpression() {
        double v = parseTerm();
        while (true) {
            skipWs();
            if (pos >= src.length()) {
                return v;
            }
            char c = src.charAt(pos);
            if (c == '+') {
                pos++;
                v = v + parseTerm();
            } else if (c == '-') {
                pos++;
                v = v - parseTerm();
            } else {
                return v;
            }
        }
    }

    private double parseTerm() {
        double v = parsePower();
        while (true) {
            skipWs();
            if (pos >= src.length()) {
                return v;
            }
            char c = src.charAt(pos);
            if (c == '*') {
                pos++;
                v = v * parsePower();
            } else if (c == '/') {
                pos++;
                v = v / parsePower();
            } else {
                return v;
            }
        }
    }

    /** power: unary ('**'|'^') power  (right associative) */
    private double parsePower() {
        double base = parseUnary();
        skipWs();
        if (pos < src.length()) {
            char c = src.charAt(pos);
            boolean isPow = false;
            if (c == '^') {
                isPow = true;
                pos++;
            } else if (c == '*') {
                if (pos + 1 < src.length() && src.charAt(pos + 1) == '*') {
                    isPow = true;
                    pos += 2;
                }
            }
            if (isPow) {
                double exp = parsePower();
                return Math.pow(base, exp);
            }
        }
        return base;
    }

    private double parseUnary() {
        skipWs();
        if (pos < src.length() && (src.charAt(pos) == '-' || src.charAt(pos) == '+')) {
            char c = src.charAt(pos);
            pos++;
            double v = parseUnary();
            return c == '-' ? -v : v;
        }
        return parsePrimary();
    }

    private double parsePrimary() {
        skipWs();
        if (pos >= src.length()) {
            return NaN;
        }
        char c = src.charAt(pos);
        if (c == '(') {
            pos++;
            double v = parseExpression();
            skipWs();
            if (pos < src.length() && src.charAt(pos) == ')') {
                pos++;
            } else {
                return NaN;
            }
            return v;
        }
        if (Character.isDigit(c) || c == '.') {
            return parseNumber();
        }
        if (Character.isLetter(c) || c == '_') {
            return parseIdent();
        }
        return NaN;
    }

    private double parseNumber() {
        int start = pos;
        boolean any = false;
        boolean exponent = false;
        boolean hasExpDigits = false;
        while (pos < src.length()) {
            char c = src.charAt(pos);
            if (Character.isDigit(c)) {
                any = true;
                if (exponent) {
                    hasExpDigits = true;
                }
                pos++;
            } else if (c == '.') {
                if (exponent) {
                    break;
                }
                any = true;
                pos++;
            } else if (c == 'e' || c == 'E') {
                if (exponent) {
                    break;
                }
                exponent = true;
                pos++;
                if (pos < src.length() && (src.charAt(pos) == '+' || src.charAt(pos) == '-')) {
                    pos++;
                }
            } else {
                break;
            }
        }
        if (!any) {
            return NaN;
        }
        String num = src.substring(start, pos);
        try {
            return Double.parseDouble(num);
        } catch (NumberFormatException e) {
            return NaN;
        }
    }

    private double parseIdent() {
        int start = pos;
        // consume a (possibly dotted) identifier: Math.pow, Math.LN2, log2, ...
        while (pos < src.length()) {
            char c = src.charAt(pos);
            if (Character.isLetterOrDigit(c) || c == '_' || c == '.') {
                pos++;
            } else {
                break;
            }
        }
        String ident = src.substring(start, pos);
        skipWs();
        if (pos < src.length() && src.charAt(pos) == '(') {
            // function call, possibly Math.func / MATH.func
            String fnName = ident;
            if (ident.indexOf('.') >= 0) {
                int dot = ident.lastIndexOf('.');
                fnName = ident.substring(dot + 1);
            }
            pos++; // consume '('
            List<Double> args = new ArrayList<Double>();
            skipWs();
            if (pos < src.length() && src.charAt(pos) == ')') {
                pos++;
            } else {
                while (true) {
                    double a = parseExpression();
                    if (Double.isNaN(a)) {
                        return NaN;
                    }
                    args.add(Double.valueOf(a));
                    skipWs();
                    if (pos < src.length() && src.charAt(pos) == ',') {
                        pos++;
                    } else {
                        break;
                    }
                }
                skipWs();
                if (pos < src.length() && src.charAt(pos) == ')') {
                    pos++;
                } else {
                    return NaN;
                }
            }
            Fn fn = FUNCS.get(fnName.toLowerCase(Locale.US));
            if (fn == null) {
                return NaN;
            }
            double[] arr = new double[args.size()];
            for (int i = 0; i < arr.length; i++) {
                arr[i] = args.get(i).doubleValue();
            }
            try {
                return fn.apply(arr);
            } catch (Exception e) {
                return NaN;
            }
        }
        // constant
        String constName = ident;
        if (ident.indexOf('.') >= 0) {
            int dot = ident.lastIndexOf('.');
            constName = ident.substring(dot + 1);
        }
        Double cv = CONSTS.get(constName.toUpperCase(Locale.US));
        if (cv != null) {
            return cv.doubleValue();
        }
        return NaN;
    }
}

File: lib/domain/tuning/math_eval.dart
import 'dart:math' as math;

/// Evaluates the arithmetic expressions accepted by tuning configuration files.
///
/// Expressions never execute code. Invalid syntax produces [double.nan], as in
/// the original Android parser. Powers associate to the right; unary signs bind
/// before powers to preserve that parser's behavior.
abstract final class MathEval {
  static double eval(String? expression) {
    if (expression == null || expression.length > 8192) return double.nan;
    final source = expression.trim();
    if (source.isEmpty) return double.nan;
    try {
      final parser = _ExpressionParser(source);
      final result = parser.expression();
      parser.skipWhitespace();
      return parser.position == source.length ? result : double.nan;
    } on FormatException {
      return double.nan;
    } on RangeError {
      return double.nan;
    } on UnsupportedError {
      return double.nan;
    }
  }
}

class _ExpressionParser {
  _ExpressionParser(this.source);

  final String source;
  int position = 0;
  int _depth = 0;
  static final _number = RegExp(r'(?:\d+(?:\.\d*)?|\.\d+)(?:[eE][+-]?\d+)?');
  static final _identifier = RegExp(r'[a-zA-Z_][a-zA-Z_0-9.]*');

  void skipWhitespace() {
    while (position < source.length && ' \t\r\n'.contains(source[position])) {
      position++;
    }
  }

  bool _consume(String token) {
    skipWhitespace();
    if (!source.startsWith(token, position)) return false;
    position += token.length;
    return true;
  }

  double expression() {
    var value = _term();
    while (true) {
      if (_consume('+')) {
        value += _term();
      } else if (_consume('-')) {
        value -= _term();
      } else {
        return value;
      }
    }
  }

  double _term() {
    var value = _power();
    while (true) {
      if (_consume('*')) {
        value *= _power();
      } else if (_consume('/')) {
        value /= _power();
      } else {
        return value;
      }
    }
  }

  double _power() {
    _enter();
    try {
      final value = _unary();
      if (_consume('**') || _consume('^')) {
        return math.pow(value, _power()).toDouble();
      }
      return value;
    } finally {
      _depth--;
    }
  }

  double _unary() {
    _enter();
    try {
      if (_consume('-')) return -_unary();
      if (_consume('+')) return _unary();
      return _primary();
    } finally {
      _depth--;
    }
  }

  void _enter() {
    if (++_depth > 128) {
      throw const FormatException('Expression nesting is too deep');
    }
  }

  double _primary() {
    if (_consume('(')) {
      final value = expression();
      if (!_consume(')')) {
        throw const FormatException('Missing closing bracket');
      }
      return value;
    }
    skipWhitespace();
    final number = _number.matchAsPrefix(source, position);
    if (number != null) {
      position = number.end;
      return double.parse(number.group(0)!);
    }
    final identifier = _identifier.matchAsPrefix(source, position);
    if (identifier == null) throw const FormatException('Expected a value');
    position = identifier.end;
    final name = identifier.group(0)!.split('.').last.toLowerCase();
    if (!_consume('(')) return _constant(name);
    final arguments = <double>[];
    if (!_consume(')')) {
      do {
        final argument = expression();
        if (argument.isNaN) return double.nan;
        arguments.add(argument);
      } while (_consume(','));
      if (!_consume(')')) {
        throw const FormatException('Missing closing bracket');
      }
    }
    return _function(name, arguments);
  }

  static double _constant(String name) => switch (name) {
    'pi' => math.pi,
    'e' => math.e,
    'ln2' => math.ln2,
    'ln10' => math.ln10,
    'log2e' => 1 / math.ln2,
    'log10e' => 1 / math.ln10,
    'sqrt2' => math.sqrt2,
    'sqrt1_2' => 1 / math.sqrt2,
    _ => double.nan,
  };

  static double _function(String name, List<double> values) {
    // Fixed arity matches the old evaluator: extra arguments are ignored and
    // missing arguments are invalid rather than silently defaulting to zero.
    if (values.isEmpty) return double.nan;
    final a = values[0];
    if (const {'pow', 'min', 'max', 'atan2', 'hypot'}.contains(name)) {
      if (values.length < 2) return double.nan;
      final b = values[1];
      return switch (name) {
        'pow' => math.pow(a, b).toDouble(),
        'min' => math.min(a, b),
        'max' => math.max(a, b),
        'atan2' => math.atan2(a, b),
        'hypot' => _hypot(a, b),
        _ => double.nan,
      };
    }
    return switch (name) {
      'log' || 'ln' => math.log(a),
      'log2' => math.log(a) / math.ln2,
      'log10' => math.log(a) / math.ln10,
      'exp' => math.exp(a),
      'sqrt' => math.sqrt(a),
      'cbrt' => a.sign * math.pow(a.abs(), 1 / 3),
      'abs' => a.abs(),
      'floor' => a.floorToDouble(),
      'ceil' => a.ceilToDouble(),
      'round' => a.isFinite ? (a + 0.5).floorToDouble() : a,
      'trunc' => a.truncateToDouble(),
      'sign' => a.sign,
      'sin' => math.sin(a),
      'cos' => math.cos(a),
      'tan' => math.tan(a),
      'asin' => math.asin(a),
      'acos' => math.acos(a),
      'atan' => math.atan(a),
      'sinh' => (math.exp(a) - math.exp(-a)) / 2,
      'cosh' => (math.exp(a) + math.exp(-a)) / 2,
      'tanh' =>
        a > 20
            ? 1
            : (a < -20 ? -1 : (math.exp(2 * a) - 1) / (math.exp(2 * a) + 1)),
      _ => double.nan,
    };
  }

  static double _hypot(double a, double b) {
    final large = math.max(a.abs(), b.abs());
    final small = math.min(a.abs(), b.abs());
    if (large.isInfinite || large == 0) return large;
    return large * math.sqrt(1 + math.pow(small / large, 2));
  }
}
